pkg/metrics/collector.go

```diff
@@ -96,21 +102,266 @@ func (c *metricsCollector) publishVolumeStats(ctx context.Context, volume *types
+		prometheus.NewDesc(
+			prometheus.BuildFQName(consts.AppName, "stats", "drive_write_latency_seconds"),
+			"Drive Write Latency",
+			[]string{"drive"}, nil),
+		prometheus.GaugeValue,
+		float64(driveStat.WriteTicks)/1000, drive.Name,
+	)
+
+	// Drive Read/Write Throughput
+	ch <- prometheus.MustNewConstMetric(
+		prometheus.NewDesc(
+			prometheus.BuildFQName(consts.AppName, "stats", "drive_read_throughput_bytes_per_second"),
+			"Drive Read Throughput",
+			[]string{"drive"}, nil),
+		prometheus.GaugeValue,
+		float64(driveStat.ReadSectors)*float64(sectorSizeBytes)*1000/float64(driveStat.ReadTicks), drive.Name,
+	)
+
+	ch <- prometheus.MustNewConstMetric(
+		prometheus.NewDesc(
+			prometheus.BuildFQName(consts.AppName, "stats", "drive_write_throughput_bytes_per_second"),
+			"Drive Write Throughput",
+			[]string{"drive"}, nil),
+		prometheus.GaugeValue,
+		float64(driveStat.WriteSectors)*float64(sectorSizeBytes)*1000/float64(driveStat.WriteTicks), drive.Name,
+	)
+
+	// Wait Time
+	ch <- prometheus.MustNewConstMetric(
+		prometheus.NewDesc(
+			prometheus.BuildFQName(consts.AppName, "stats", "drive_wait_time_seconds"),
+			"Drive Wait Time",
+			[]string{"drive"}, nil),
+		prometheus.GaugeValue,
+		float64(driveStat.TimeInQueue)/1000, drive.Name,
+	)
+}
+
+func readDriveStats(filePath string) (*driveStats, error) {
+	file, err := os.Open(filePath)
+	if err != nil {
+		return nil, fmt.Errorf("opening stat file: %w", err)
+	}
+	defer file.Close()
+
+	return parseStats(file)
+}
+
+func parseStats(reader *os.File) (*driveStats, error) {
+	scanner := bufio.NewScanner(reader)
+	if scanner.Scan() {
+		fields := strings.Fields(scanner.Text())
+		if len(fields) < 11 {
+			return nil, fmt.Errorf("unexpected format in stat file")
+		}
+
+		return parseDriveStats(fields)
+	}
+
+	if err := scanner.Err(); err != nil {
+		return nil, fmt.Errorf("reading stat file: %w", err)
+	}
+
+	return nil, fmt.Errorf("stat file is empty")
+}
+
+func parseDriveStats(fields []string) (*driveStats, error) {
+	var stats driveStats
+	var err error
+
+	stats.ReadSectors, err = strconv.ParseInt(fields[2], 10, 64)
+	if err != nil {
+		return nil, fmt.Errorf("parsing ReadSectors: %w", err)
+	}
+	stats.ReadTicks, err = strconv.ParseInt(fields[3], 10, 64)
+	if err != nil {
+		return nil, fmt.Errorf("parsing ReadTicks: %w", err)
+	}
+	stats.WriteSectors, err = strconv.ParseInt(fields[6], 10, 64)
+	if err != nil {
+		return nil, fmt.Errorf("parsing WriteSectors: %w", err)
+	}
+	stats.WriteTicks, err = strconv.ParseInt(fields[7], 10, 64)
+	if err != nil {
+		return nil, fmt.Errorf("parsing WriteTicks: %w", err)
+	}
+	stats.TimeInQueue, err = strconv.ParseInt(fields[9], 10, 64)
+	if err != nil {
+		return nil, fmt.Errorf("parsing TimeInQueue: %w", err)
+	}
+
+	return &stats, nil
+}
+
+func getSectorSize(device string) (int64, error) {
+	// Construct the file path to the sector size file
+	filePath := "/sys/block/" + device + "/queue/hw_sector_size"
```

Comment:
1. This file doesn't exist for disk partitions. For partitions, find it is disk and fetch sector size from it and use it.
2. if `/sys/class/block/<device>/queue/hw_sector_size` is missing for disk, use 512 as default block size.
3. Move this function as `GetHardwareSectorSize` to `pkg/device/sysfs_linux.go`

Reply:
I found a `defaultBlockSize` as 512 in the `sysfs_linux.go`, Since it's the same value, should I go ahead and use it? but I'm thinking block size and sector size aren't the same thing, I could create a new constant just for sector size.

Reply:
Most of the system have default block/sector size as 512.  you could define default sector size as default block size.